src/elf.c

```diff
@@ -342,13 +342,12 @@ bool elf_load(elf_t *e, riscv_t *rv, memory_t *mem)
         /* memcpy required range */
         const int to_copy = min(phdr->p_memsz, phdr->p_filesz);
         if (to_copy)
-            memory_write(mem, phdr->p_vaddr, e->raw_data + phdr->p_offset,
-                         to_copy);
+            memory_write(phdr->p_vaddr, e->raw_data + phdr->p_offset, to_copy);
```

Comment:
Don't do that. For public C functions like `memory_write`, we have to ensure the encapsulation (or information hiding) of internal states. This means that we may create several RISC-V virtual machine instances simultaneously, and there must be appropriate responses for the corresponding calls. Keep this design principle in mind.

Reply:
Does this mean we still need to pass the virtual machine instances for all read and write functions, including read and write instructions?

Reply:
Standalone functions such as `memory_write` should come with a parameter as context. However, the function pointers inside a structure can omit the context.